libr/core/disasm.c

```diff
@@ -6583,3 +6583,186 @@ R_API int r_core_disasm_pdi(RCore *core, int nb_opcodes, int nb_bytes, int fmt)
+		}
+	}
+	const ut64 ocoffset = core->offset;
+	const ut64 ocbsize = core->blocksize;
+	RAnalEsil *esil = core->anal->esil;
+	RReg *reg = core->anal->reg;
+	RList *ocache = core->io->cache;
+	RCache *ocacheb = core->io->buffer;
+	const int ocached = core->io->cached;
+	const bool ocachec = r_config_get_i (core->config, "io.cache");
+	const bool ojmp_lines = r_config_get_i (core->config, "asm.lines");
+	const bool hide_branches = r_config_get_i (core->config, "asm.followcf.hide");
+	if (ocacheb && ocache) {
+		RCache *c = R_NEW0 (RCache);
+		c->base = ocacheb->base;
+		if (ocacheb->len) {
+			ut8 *buf = malloc (ocacheb->len);
+			if (!buf) {
+				pj_free (pj);
+				r_cache_free (c);
+				return false;
+			}
+			memcpy (buf, ocacheb->buf, ocacheb->len);
+			c->buf = buf;
+			c->len = ocacheb->len;
+		}
+		core->io->buffer = c;
+		core->io->cache = r_list_clone (ocache);
+	} else {
+		r_io_cache_init (core->io);
+	}
+	r_reg_arena_push (reg);
+	r_config_set_i (core->config, "io.cache", 1);
+	r_config_set_i (core->config, "asm.lines", 0);
+	const ut64 read_len = 32;
+	size_t buf_sz = 0x100, block_sz = 0, block_instr = 0;
+	RRegItem *pc = r_reg_get (reg, "PC", R_REG_TYPE_ALL);
+	ut64 block_start = r_reg_get_value (reg, pc);
+	ut8 *buf = malloc (buf_sz);
+	int i;
+	for (i = 0; i < nb_opcodes; i++) {
+		const ut64 op_addr = r_reg_get_value (reg, pc);
+		if (block_sz + read_len > buf_sz) {
+			buf_sz *= 2;
+			ut8 *tmp = realloc (buf, buf_sz);
+			if (!tmp) {
+				break;
+			}
+			buf = tmp;
+		}
+		if (!r_io_read_at_mapped (core->io, op_addr, buf + block_sz, read_len)) {
+			break;
+		}
+		RAnalOp op;
+		int ret = r_anal_op (core->anal, &op, op_addr, buf + block_sz, read_len, R_ANAL_OP_MASK_ESIL);
+		const bool invalid_instr = ret < 1 || op.size < 1;
+		bool end_of_block = false;
+		switch (op.type & R_ANAL_OP_TYPE_MASK & ~R_ANAL_OP_HINT_MASK) {
+		case R_ANAL_OP_TYPE_JMP:
+		case R_ANAL_OP_TYPE_UJMP:
+		case R_ANAL_OP_TYPE_CALL:
+		case R_ANAL_OP_TYPE_UCALL:
+		case R_ANAL_OP_TYPE_RET:
+			end_of_block = true;
+			if (hide_branches) {
+				i--;
+			} else {
+				block_instr++;
+				block_sz += op.size;
+			}
+			break;
+		default:
+			if (i + 1 >= nb_opcodes || invalid_instr) {
+				end_of_block = true;
+			}
+			block_instr++;
+			block_sz += op.size;
+			break;
+		}
+		if (end_of_block) {
+		print_delay_slots:
+			if (block_instr) {
+				switch (mode) {
+				case R_MODE_JSON:
+					r_core_print_disasm_json (core, block_start, buf, block_sz, block_instr, pj);
+					break;
+				case R_MODE_SIMPLE:
+					r_core_seek (core, block_start, true);
+					r_core_disasm_pdi (core, block_instr, block_sz, 0);
+					break;
+				case R_MODE_SIMPLEST:
+					r_core_seek (core, block_start, true);
+					r_core_print_disasm_instructions (core, block_sz, block_instr);
+					break;
+				default:
+					r_core_print_disasm (core->print, core, block_start, buf, block_sz, block_instr, 0, block_sz, false, NULL, NULL);
+					break;
+				}
+			}
+			if (op.delay) {
```

Comment:
i assume you have a testcase for MIPS to demonstrate the delay slot emulation thing here